store/postgres/src/mapping.rs

```diff
@@ -0,0 +1,1141 @@
+    pub fn column_for_field(&self, field: &str) -> Result<&Column, StoreError> {
+        self.columns
+            .iter()
+            .find(|column| &column.field == field)
+            .ok_or_else(|| StoreError::UnknownField(field.to_string()))
+    }
+
+    pub fn primary_key(&self) -> Reference {
+        Reference {
+            table: self.name.clone(),
+            column: PRIMARY_KEY_COLUMN.into(),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn reference(&self, name: &SqlName) -> Result<Reference, StoreError> {
+        let attr = self.column(name)?;
+        Ok(Reference {
+            table: self.name.clone(),
+            column: attr.name.clone(),
+        })
+    }
+
+    pub fn collection_name(gql_type_name: &str) -> SqlName {
+        SqlName::from(gql_type_name.to_snake_case().to_plural())
+    }
+
+    pub fn object_name(gql_type_name: &str) -> String {
+        gql_type_name.to_snake_case()
+    }
+}
+
+impl AsDdl for Table {
+    fn fmt(&self, f: &mut dyn fmt::Write, mapping: &Mapping) -> fmt::Result {
+        write!(f, "create table {}.{} (\n", mapping.schema, self.name)?;
+        for column in self.columns.iter().filter(|col| !col.is_derived()) {
+            write!(f, "    ")?;
+            column.fmt(f, mapping)?;
+            write!(f, ",\n")?;
+        }
+        // Add block_range column and constraint
+        write!(
+            f,
+            "\n        {block_range}          int4range not null,
+        exclude using gist   (id with =, {block_range} with &&)\n",
+            block_range = BLOCK_RANGE
+        )?;
+        if self.columns.iter().any(|col| col.is_derived()) {
+            write!(f, "\n     -- derived fields (not stored in this table)\n")?;
+            for column in self.columns.iter().filter(|col| col.is_derived()) {
+                column.fmt(f, mapping)?;
+                for reference in &column.references {
+                    write!(f, " references {}({})", reference.table, reference.column)?;
+                }
+                write!(f, "\n")?;
+            }
+        }
+        write!(f, ");\n")?;
+
+        for (i, column) in self
+            .columns
+            .iter()
+            .filter(|col| !col.is_derived())
+            .enumerate()
+        {
+            if !column.is_primary_key()
+                && !column.is_reference()
+                && !column.is_list()
+                && column.column_type == ColumnType::String
+            {
+                write!(
+                f,
+                "create index attr_{table_index}_{column_index}_{table_name}_{column_name}\n    on {schema_name}.{table_name} using btree(left({column_name}, {prefix_size}));\n",
+                table_index=self.position,
+                table_name=self.name,
+                column_index=i + 1,
+                column_name=column.name,
+                schema_name=mapping.schema,
+                prefix_size=STRING_PREFIX_SIZE,
+            )?;
```

Comment:
Can `table_name` cause to hit the PG identifier length limit? If so we should use only the `table_index` and not the name.

Reply:
Oh this comment https://github.com/graphprotocol/graph-node/blob/master/store/postgres/src/entities.rs#L1161 clarified things for me, can we have a comment explaining this here as well?

Reply:
With that, are you worried that we might have two table names that have the same 63 character prefix? If so, I would generate a table name as we do right now, and if it's less than 55 chars, just use it; if it is longer than that, truncate to 55 chars and tack the `table_index` onto the end. But there's a bunch of gyrations when I resolve references needed to make that possible (that relies right now on the fact that we can compute the table name from the object name alone)

Reply:
We've decided to let this be, if it's a bug or by design we can decide later.

Reply:
Still, can we have comments noting this somewhere near the code that sets the table name?

Reply:
Left a comment on `SqlName` as part of 68ebd99d